### DISCUSSION TITLE
Expression syntax in Conditionals

### DISCUSSION THREAD

**Discussion by bleaf-GH** [community/community#58280] [Actions] [answered]
### Select Topic Area

Question

### Body

Hello! I'd like to report some unexpected behavior and see if anyone has some advice on next steps.

Context: using `dorny/paths-filter@v2`:
`outputs` = For each filter, it sets output variable named by the filter to the text:
`'true'` - if any changed files match any filter rules
`'false'` - if no changed files match any filter rules

`outputs.changes` = JSON array with names of all filters matching any of the changed files.

Expected Outcome:
```
- name: Echo (Testing)
  run: |
    echo "Unpackaged Outputs - No Changes: ${{ steps.unpackaged_changes.outputs.unpackaged }}"
    echo "Packaged Outputs Test Array - No Changes: ${{ steps.packaged_changes.outputs.changes }}"
    echo "Packaged Outputs Test fromJson - No Changes: ${{ fromJson(steps.packaged_changes.outputs.changes)[0] }}"

Unpackaged Outputs - No Changes: false
Packaged Outputs Test Array - No Changes: []
Packaged Outputs Test fromJson - No Changes: 

- name: Test Failure
  if: steps.unpackaged_changes.outputs.unpackaged == 'true' || fromJson(steps.packaged_changes.outputs.changes)[0] != ''
  run: | 
    echo "FAILURE"
    exit 1

- name: Test Success
  run: echo "SUCCESS"

SUCCESS
```

However, when I was asked to standardize the conditional to match the syntax used in other .yml files in the repository:
```
if: ${{ steps.unpackaged_changes.outputs.unpackaged == 'true' }} || ${{ fromJson(steps.packaged_changes.outputs.changes)[0] != '' }}
FAILURE
```

When I apply the expression syntax to just the first expression:
```
if: ${{ steps.unpackaged_changes.outputs.unpackaged == 'true' }} || fromJson(steps.packaged_changes.outputs.changes)[0] != ''
FAILURE

if: ${{ steps.unpackaged_changes.outputs.unpackaged }} == 'true' || fromJson(steps.packaged_changes.outputs.changes)[0] != ''
FAILURE
```

when I remove it:
```
if: steps.unpackaged_changes.outputs.unpackaged == 'true' || fromJson(steps.packaged_changes.outputs.changes)[0] != ''
...
SUCCESS
```

The reverse case involving changes to a file has the same issue: the conditional isn't evaluating as it should with Expression syntax applied, and seems to evaluate properly when the `${{}}` syntax is removed. I understand that Github docs state that this syntax isn't required for `if` conditionals, but I didn't realize it would actually change how the code is interpreted. Does anyone have any insight here? Thanks for taking a look!

**Comment by airtower-luna** [comment, marked as answer]
If you use `${{ ... }}` in `if` expressions there's a pitfall: Everything *not* inside the brackets is interpreted as a *literal string*, and the evaluation results are substituted. The result will be a non-empty string, and a non-empty string evaluates to `true`.

I'd strongly recommend writing your `if` expressions without explicit `${{ ... }}`, that way the runner will implicitly add them around the whole expression. If you absolutely want to write them you have to wrap them around the whole expression anyway, which is redundant. :wink:
